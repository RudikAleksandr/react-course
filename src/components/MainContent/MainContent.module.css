.mainContent {
  min-height: calc(100vh - 478px);
  background-color: #232323;
}

.filterSortRow {
  border-bottom: 3px solid #424242;
  box-shadow: 0 1px 0 black;
}
