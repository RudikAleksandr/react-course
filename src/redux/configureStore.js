import { configureStore as configureStoreToolkit } from '@reduxjs/toolkit';
import thunkMiddleware from 'redux-thunk';
import moviesReducer from './moviesSlice';

const configureStore = (preloadedState) => (
  configureStoreToolkit({
    preloadedState,
    reducer: {
      movies: moviesReducer,
    },
    middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(thunkMiddleware),
  })
);

export default configureStore;
