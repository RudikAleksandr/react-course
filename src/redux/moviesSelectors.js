
const selectListMovies = (state) => state.movies.moviesOrder.map(id => state.movies.movies[id]);

const selectSearch = (state) => state.movies.search;

const selectActiveMovieGenre = (state) => state.movies.activeMovieGenre;

const selectActiveSortType = (state) => state.movies.activeSortType;

export {
  selectListMovies,
  selectSearch,
  selectActiveMovieGenre,
  selectActiveSortType
};