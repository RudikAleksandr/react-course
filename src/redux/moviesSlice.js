import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import * as moviesAPI from '../api/moviesAPI';
import { groupByKey } from '../utils/grouping';

export const initialState = {
  movies: {},
  moviesOrder: [],
  search: '',
  activeMovieGenre: '',
  activeSortType: 'release_date'
};

export const getListMovies = createAsyncThunk(
  'movies/getListMovies',
  moviesAPI.getListMovies
);

export const getMovie = createAsyncThunk(
  'movies/getMovie',
  moviesAPI.getMovie
);

export const updateMovie = createAsyncThunk(
  'movies/updateMovie',
  moviesAPI.updateMovie
);

export const createMovie = createAsyncThunk(
  'movies/createMovie',
  moviesAPI.createMovie
);

export const deleteMovie = createAsyncThunk(
  'movies/deleteMovie',
  moviesAPI.deleteMovie
);

const moviesSlice = createSlice({
  name: 'movies',
  initialState,
  reducers: {
    setSearch: (state, { payload }) => {
      state.search = payload;
    },
    setActiveMovieGenre: (state, { payload }) => {
      state.activeMovieGenre = payload;
    },
    setActiveSortType: (state, { payload }) => {
      state.activeSortType = payload;
    }
  },
  extraReducers: {
    [getListMovies.fulfilled]: (state, { payload }) => {
      state.movies = groupByKey('id', payload.data);
      state.moviesOrder = payload.data.map((movie) => movie.id);
    },
    [updateMovie.fulfilled]: (state, { payload }) => {
      state.movies[payload.id] = payload;
    },
  },
});

export const {
  setSearch,
  setActiveMovieGenre,
  setActiveSortType
} = moviesSlice.actions;

export default moviesSlice.reducer;