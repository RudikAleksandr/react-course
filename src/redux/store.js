import { configureStore } from '@reduxjs/toolkit';
import moviesReducer from './moviesSlice';
import thunkMiddleware from 'redux-thunk'

const store = configureStore({
  reducer: {
    movies: moviesReducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(thunkMiddleware),
});

export default store;