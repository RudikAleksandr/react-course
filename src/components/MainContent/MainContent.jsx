import React, { useCallback } from 'react';
import { Row, Col } from 'react-bootstrap';
import FilterMovies from '../FilterMovies';
import SortMovies from '../SortMovies';
import ListMovies from '../ListMovies';
import { useDispatch, useSelector } from 'react-redux';
import { selectActiveMovieGenre, selectActiveSortType, selectSearch } from '../../redux/moviesSelectors';
import { setActiveMovieGenre, setActiveSortType } from '../../redux/moviesSlice';
import cn from './MainContent.module.css';

const MainContent = () => {
  const activeMovieGenre = useSelector(selectActiveMovieGenre);
  const activeSortType = useSelector(selectActiveSortType);
  const search = useSelector(selectSearch);
  const dispatch = useDispatch();

  const handleClickMovieGenre = useCallback((movieGenre) => {
    dispatch(setActiveMovieGenre(movieGenre));
  }, [dispatch])

  const handleSelectSortType = useCallback((sortType) => {
    dispatch(setActiveSortType(sortType));
  }, [dispatch])

  return (
    <Col xs="12" lg="11" xl="9">
      <Row className={`flex-column align-items-center mt-2 ${cn.mainContent}`}>
        <Col className="mt-1" xs="11">
          <Row className={cn.filterSortRow}>
            <Col xs="8">
              <FilterMovies
                activeType={activeMovieGenre}
                onClickMovieGenre={handleClickMovieGenre}
              />
            </Col>
            <Col xs="4">
              <SortMovies
                activeType={activeSortType}
                onSelectSortType={handleSelectSortType}
              />
            </Col>
          </Row>
          <ListMovies
            search={search}
            sortBy={activeSortType}
            genre={activeMovieGenre}
            sortType={activeSortType}
          />
        </Col>
      </Row>
    </Col>
  )
};

export default MainContent;
