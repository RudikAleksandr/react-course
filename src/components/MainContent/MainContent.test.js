
import { render, } from '@testing-library/react';
import MainContent from './MainContent';
import * as reactRedux from 'react-redux';
import * as React from 'react';
import configureStore from 'redux-mock-store'
import { initialState } from '../../redux/moviesSlice';
import thunk from 'redux-thunk'

const getComponent = (store) => (
  <reactRedux.Provider store={store}>
    <MainContent />
  </reactRedux.Provider>
);

describe('MainContent component', () => {
  test('render MainContent snapshot', () => {
    const store = configureStore([thunk])({ movies: initialState });
    const component = getComponent(store);
    const { asFragment } = render(component);

    expect(asFragment(component)).toMatchSnapshot();
  });
});
