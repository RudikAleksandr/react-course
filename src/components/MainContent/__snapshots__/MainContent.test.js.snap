// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainContent component render MainContent snapshot 1`] = `
<DocumentFragment>
  <div
    class="col-xl-9 col-lg-11 col-12"
  >
    <div
      class="flex-column align-items-center mt-2 mainContent row"
    >
      <div
        class="mt-1 col-11"
      >
        <div
          class="filterSortRow row"
        >
          <div
            class="col-8"
          >
            <div
              class="filterMovies row"
            >
              <div
                class="pl-0 col"
              >
                <button
                  class="filterButton  btn"
                  type="button"
                >
                  all
                </button>
                <button
                  class="filterButton  btn"
                  type="button"
                >
                  documentary
                </button>
                <button
                  class="filterButton  btn"
                  type="button"
                >
                  comedy
                </button>
                <button
                  class="filterButton  btn"
                  type="button"
                >
                  horror
                </button>
                <button
                  class="filterButton  btn"
                  type="button"
                >
                  crime
                </button>
              </div>
            </div>
          </div>
          <div
            class="col-4"
          >
            <div
              class="align-items-center justify-content-end sortMovies row"
            >
              <span
                class="sortByLabel"
              >
                sort by
              </span>
              <div
                class="sortTypeDropdown dropdown btn-group"
                role="group"
              >
                <button
                  aria-expanded="false"
                  aria-haspopup="true"
                  class="dropdown-toggle btn"
                  type="button"
                >
                  release date
                </button>
              </div>
            </div>
          </div>
        </div>
        <div
          class="flex-column mb-4 listMoviesContainer row"
        >
          <div
            class="pl-0 mt-4 mb-4 col"
          >
            <span
              class="countMovies"
            >
              0
            </span>
            <span
              class="moviesFoundLabel"
            >
              movies found
            </span>
          </div>
          <div
            class="col"
          >
            <div
              class="noMovieFound"
            >
              <h2
                class="noMovieFoundLabel"
              >
                No Movie Found
              </h2>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
